### PR TITLE
Node json should be optional

### PR DESCRIPTION
**Pull request by duncangrant**
The Node element of the container json is optional as well as nullable.

In PR #277 I tested against docker-engine provisioned through docker-machine and docker-swarm but not against vanilla docker engine.  I have now tested against all three.

### CODE REVIEW

**Comment by grkvlt** [comment]
@duncangrant Thanks, good catch! The `Optional<?>` type in autovalue is a nice feature!

**Comment by grkvlt** [comment]
@andreaturli This is _another_ Swarm related patch (for master and 1.9.x) that fixes an issue we missed previously, and @duncangrant has now added some tests, too. Can you take a look?

File: docker/src/main/java/org/jclouds/docker/domain/Container.java [hunk 1]
```diff
@@ -77,7 +78,7 @@
 
    @Nullable public abstract String processLabel();
 
-   @Nullable public abstract Node node();
+   @Nullable public abstract Optional<Node> node();
```
Comment (nacx) [inline docker/src/main/java/org/jclouds/docker/domain/Container.java]: Optional fields should never be nullable.

File: docker/src/main/java/org/jclouds/docker/domain/Container.java [hunk 2]
```diff
@@ -93,7 +94,7 @@ public static Container create(String id, Date created, String path, String name
                                   String resolvConfPath, Map<String, String> volumes, HostConfig hostConfig,
                                   String driver, String execDriver, Map<String, Boolean> volumesRW, String command,
                                   String status, List<Port> ports, String hostnamePath, String hostsPath,
-                                  String mountLabel, String processLabel, Node node) {
+                                  String mountLabel, String processLabel, Optional<Node> node) {
       return new AutoValue_Container(id, created, path, name, copyOf(args), config, state, image, networkSettings,
               sysInitPath, resolvConfPath, copyOf(volumes), hostConfig, driver, execDriver, copyOf(volumesRW), command,
               status, copyOf(ports), hostnamePath, hostsPath, mountLabel, processLabel, node);
```
Comment (nacx) [inline docker/src/main/java/org/jclouds/docker/domain/Container.java]: Use `Optional.fromNullable(node)` to set the node.

File: docker/src/main/java/org/jclouds/docker/compute/functions/ContainerToNodeMetadata.java [hunk 3]
```diff
@@ -113,8 +112,8 @@ private String cleanUpName(String name) {
 
    private List<String> getPublicIpAddresses(Container container) {
       String dockerIpAddress;
-      if (container.node() != null && !Strings.isNullOrEmpty(container.node().ip())) {
-         dockerIpAddress = container.node().ip();
+      if (container.node() != null && container.node().isPresent()) {
```
Comment (nacx) [inline docker/src/main/java/org/jclouds/docker/compute/functions/ContainerToNodeMetadata.java]: If using optionals we should get rid of the null check and assume it is never null.

**Comment by duncangrant** [comment]
@grkvlt @nacx Thanks for you comments.  I think I've addressed them.

File: docker/src/main/java/org/jclouds/docker/domain/Container.java [hunk 4]
```diff
@@ -266,7 +267,7 @@ public Builder fromContainer(Container in) {
                  .sysInitPath(in.sysInitPath()).resolvConfPath(in.resolvConfPath()).driver(in.driver())
                  .execDriver(in.execDriver()).volumes(in.volumes()).hostConfig(in.hostConfig()).volumesRW(in.volumesRW())
                  .command(in.command()).status(in.status()).ports(in.ports()).hostnamePath(in.hostnamePath())
-                 .hostsPath(in.hostsPath()).mountLabel(in.mountLabel()).processLabel(in.processLabel()).node(in.node());
+                 .hostsPath(in.hostsPath()).mountLabel(in.mountLabel()).processLabel(in.processLabel()).node(in.node().get());
```
Comment (grkvlt) [inline docker/src/main/java/org/jclouds/docker/domain/Container.java]: This will fail when `node` is `Optional.absent()`

**Comment by grkvlt** [comment]
@duncangrant LGTM @nacx can you take another look?

File: docker/src/main/java/org/jclouds/docker/compute/functions/ContainerToNodeMetadata.java (hunk 3 above)
Comment (nacx) [inline docker/src/main/java/org/jclouds/docker/compute/functions/ContainerToNodeMetadata.java]: Remove the `!= null` check as it will never be `null`.

**Comment by nacx** [comment]
Just one last minor comment. Let's address this and squash directly so we can cleanly merge the PR. Thanks @duncangrant @grkvlt!

**Comment by duncangrant** [comment]
Addressed and merged.

thanks Duncan

**Comment by nacx** [comment]
Pushed to master as [f47223b5](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/f47223b5). Thanks!

**Comment by grkvlt** [comment]
@duncangrant can you also make a PR for the 1.9.x branch, please?

**Pull request closed without merging**